internal/pkg/apply/apply.go

```diff
@@ -43,89 +48,294 @@ type Apply struct {
 
 	// Commit is a git commit object
 	Commit *object.Commit
+
+	// Flag; deletes objects previously applied, but not applied currently.
+	Prune bool
+
+	// Flag; don't actually persist created/updated/deleted resources.
+	DryRun bool
+
+	// List of previous inventory objects associated with director.
+	PrevInventory []unstructured.Unstructured
+
+	// List of resources managed by previous inventory objects.
+	PrevResources map[resid.ResId]bool
 }
 
 // Result contains the Apply Result
 type Result struct {
 	Resources clik8s.ResourceConfigs
 }
 
+const InventoryId = "config.k8s.io/InventoryId"
+
 // Do executes the apply
 func (a *Apply) Do() (Result, error) {
 	fmt.Fprintf(a.Out, "Doing `cli-utils apply`\n")
 
-	// TODO(Liuijngfang1): add a dry-run for all objects
-	// When the dry-run passes, proceed to the actual apply
+	if a.DryRun {
+		fmt.Fprintf(a.Out, "\n**Dry Run...not actually persisting**\n")
+	}
+	fmt.Fprintf(a.Out, "\n")
 
+	var director *unstructured.Unstructured
+	var inventoryRef *metav1.OwnerReference
 	for _, u := range normalizeResourceOrdering(a.Resources) {
 		if util.HasAnnotation(u, inventory.ContentAnnotation) {
-			var err error
-			u, err = a.updateInventoryObject(u)
+			// Validate that i == 0
+			if director == nil {
+				var err error
+				director, err = a.getInventoryDirector(u)
+				if err != nil && !errors.IsNotFound(err) {
+					fmt.Fprintf(a.Out, "Failure during inventory director search: %v\n", err)
+					break
+				}
+				if err != nil && errors.IsNotFound(err) {
+					director, err = a.createInventoryDirector(u)
+					if err != nil {
+						fmt.Fprintf(a.Out, "Failure creating inventory director: %v\n", err)
+						break
+					}
+				}
+				a.getPrevInventory(director)
+			}
+			directorRef := createOwnerReference(director)
+			a.addOwnerReference(u, *directorRef)
+			suffix := createUniqueId()
+			u.SetName(u.GetName() + "-" + suffix)
```

Comment:
Based on the change, there is a new suffix created and appended to `u` every time the apply is called. You can use the hash annotation in the Inventory object to skip creating a new `u` when nothing is changed in the apply list.